File: contest2/f/f.go
package main

import (
	"bufio"
	"bytes"
	"errors"
	"fmt"
	"os"
	"strconv"
)

const (
	module = 7340033
	root   = 5
	invert = 4404020
	power  = 1 << 20
)

var (
	errZeroFirstCoeff = errors.New("The ears of a dead donkey")
	fabric            modularFabric
)

func ScanInt(scanner *bufio.Scanner) int {
	if !scanner.Scan() {
		panic("nothing to scan")
	}
	str := scanner.Text()
	n, err := strconv.Atoi(str)
	if err != nil {
		panic(err)
	}
	return n
}

const whitespaceSymbols = "\t\n\v\f\r "

func splitFunc(data []byte, atEOF bool) (advance int, token []byte, err error) {
	if atEOF && len(data) == 0 {
		return 0, nil, nil
	}

	i := bytes.IndexAny(data, whitespaceSymbols)
	if i > 0 {
		return i + 1, data[:i], nil
	}
	if i == -1 {
		if atEOF {
			return len(data), data, nil
		}
		return 0, nil, nil
	}

	trimmed := bytes.TrimLeft(data, whitespaceSymbols)
	advance = len(data) - len(trimmed)
	if atEOF && len(trimmed) != 0 {
		trimmed = trimmed[:bytes.IndexAny(trimmed, whitespaceSymbols)]
		return advance + len(trimmed), trimmed, nil
	}
	return advance, nil, nil
}

type modular struct {
	value  int
	module int
}

type modularFabric struct {
	module int
}

func initFabric(module int) modularFabric {
	return modularFabric{module: module}
}

func (f modularFabric) buildModular(number int) modular {
	ans := modular{value: number % f.module}
	ans.normalize()
	return ans
}

func (m *modular) normalize() {
	if m.value >= 0 {
		m.value %= m.module
		return
	}
	m.value = (m.value + m.module*(-m.value/m.module+1)) % m.module
}

func sumModulars(first modular, second modular) modular {
	ans := modular{value: first.value + second.value}
	ans.normalize()
	return ans
}

func subModulars(first modular, second modular) modular {
	ans := modular{value: first.value - second.value}
	ans.normalize()
	return ans
}

func multModulars(first modular, second modular) modular {
	ans := modular{value: first.value * second.value}
	ans.normalize()
	return ans
}

func findInvert(m modular) modular {
	_, invert := euclidCoeffs(m.module, m.value)
	ans := modular{value: invert}
	ans.normalize()
	return ans
}

func (m modular) String() string {
	return fmt.Sprintf("%d", m.value)
}

func ScanModular(scanner *bufio.Scanner, f modularFabric) modular {
	return f.buildModular(ScanInt(scanner))
}

func getReversed(num int, log int) int {
	left := 0
	right := log - 1
	for left < right {
		bitLeft := (num & (1 << left)) >> left
		bitRight := (num & (1 << right)) >> right
		num ^= (bitLeft << left)
		num ^= (bitRight << left)
		num ^= (bitRight << right)
		num ^= (bitLeft << right)
		left++
		right--
	}
	return num
}

func euclidCoeffs(a, b int) (int, int) {
	var (
		prevX = fabric.buildModular(0)
		prevY = fabric.buildModular(1)
		x     = fabric.buildModular(1)
		y     = fabric.buildModular(-(a / b))
	)
	a, b = b, a%b
	for b != 0 {
		q := fabric.buildModular(a / b)
		if a%b != 0 {
			tmpX := x
			tmpY := y
			x = subModulars(prevX, multModulars(x, q))
			y = subModulars(prevY, multModulars(y, q))
			prevX = tmpX
			prevY = tmpY
		}
		a, b = b, a%b
	}
	return x.value, y.value
}

func max(a, b int) int {
	if a < b {
		return b
	}
	return a
}

func fft(poly []int, log int, isInvertFFT bool) {
	if len(poly) == 1 {
		return
	}

	for i := 0; i < len(poly); i++ {
		reversed := getReversed(i, log)
		if i < reversed {
			poly[i], poly[reversed] = poly[reversed], poly[i]
		}
	}

	curOffset := 1
	for i := 0; i < log; i++ {
		var curRoot int
		if isInvertFFT {
			curRoot = invert
		} else {
			curRoot = root
		}
		for j := curOffset * 2; j < power; j <<= 1 {
			curRoot = (curRoot * curRoot) % module
		}
		for j := 0; j < len(poly); j += curOffset * 2 {
			curRoot := 1
			for k := 0; k < curOffset; k++ {
				tmp1 := poly[k+j]
				tmp2 := poly[j+k+curOffset]
				poly[k+j] = tmp1 + (curRoot*tmp2)%module
				if poly[k+j] > module {
					poly[k+j] -= module
				}
				poly[k+j+curOffset] = tmp1 - (curRoot*tmp2)%module
				if poly[k+j+curOffset] < 0 {
					poly[k+j+curOffset] += module
				}
				curRoot = (curRoot * curRoot) % module
			}
		}
		curOffset *= 2
	}

	if isInvertFFT {
		_, inverted := euclidCoeffs(module, len(poly))
		for i := range poly {
			poly[i] = (poly[i] * inverted) % module
		}
	}
}

func multiply(firstPoly []int, secondPolyArg []int) []int {
	var (
		secondPoly []int
	)
	secondPoly = append(secondPoly, secondPolyArg...)

	if len(firstPoly) == 0 || len(secondPoly) == 0 {
		firstPoly = []int{0}
		return firstPoly
	}
	for len(firstPoly) > 1 && firstPoly[len(firstPoly)-1] == 0 {
		firstPoly = firstPoly[:(len(firstPoly) - 1)]
	}
	for len(secondPoly) > 1 && secondPoly[len(secondPoly)-1] == 0 {
		secondPoly = secondPoly[:len(secondPoly)-1]
	}
	if firstPoly[len(firstPoly)-1] == 0 || secondPoly[len(secondPoly)-1] == 0 {
		firstPoly = []int{0}
		return firstPoly
	}
	for len(firstPoly) < len(secondPoly) {
		firstPoly = append(firstPoly, 0)
	}

	for len(secondPoly) < len(firstPoly) {
		secondPoly = append(secondPoly, 0)
	}

	minDegTwo := 1
	log := 0
	for minDegTwo < len(firstPoly) {
		log++
		minDegTwo <<= 1
	}
	minDegTwo <<= 1
	log++

	for len(firstPoly) < minDegTwo {
		firstPoly = append(firstPoly, 0)
		secondPoly = append(secondPoly, 0)
	}

	fft(firstPoly, log, false)
	fft(secondPoly, log, false)

	for i := range firstPoly {
		firstPoly[i] = (firstPoly[i] * secondPoly[i]) % module
	}

	fft(firstPoly, log, true)
	for len(firstPoly) > 0 && firstPoly[len(firstPoly)-1] == 0 {
		firstPoly = firstPoly[:(len(firstPoly) - 1)]
	}
	return firstPoly
}

func findInvertPoly(poly []int, deg int, fabric *modularFabric) ([]int, error) {
	if poly[0] == 0 {
		return nil, errZeroFirstCoeff
	}
	for len(poly) < deg {
		extra := make([]int, deg-len(poly))
		poly = append(poly, extra...)
	}
	var (
		invert = make([]int, len(poly))
		curDeg = 1
	)

	_, invert[0] = euclidCoeffs(module, poly[0])
	for curDeg < deg {
		firstPart := make([]int, 0)
		for i := 0; i < curDeg; i++ {
			firstPart = append(firstPart, poly[i])
		}
		secondPart := make([]int, 0)
		for i := curDeg; i < len(poly); i++ {
			secondPart = append(secondPart, poly[i])
		}

		firstPart = multiply(firstPart, invert)
		if len(firstPart) > deg {
			firstPart = firstPart[:deg+1]
		}
		if curDeg >= len(firstPart) {
			firstPart = make([]int, 0)
		} else {
			firstPart = firstPart[curDeg:]
		}
		secondPart = multiply(secondPart, invert)
		if len(secondPart) > deg {
			secondPart = secondPart[:deg+1]
		}
		for i := 0; i < max(len(firstPart), len(secondPart)); i++ {
			if i == len(firstPart) {
				firstPart = append(firstPart, 0)
			}
			if i < len(firstPart) && i < len(secondPart) {
				firstPart[i] = (-firstPart[i] - secondPart[i] + 2*module) % module
			} else if i < len(secondPart) {
				firstPart[i] = (-secondPart[i] + module) % module
			} else {
				firstPart[i] = (-firstPart[i] + module) % module
			}
		}
		firstPart = multiply(firstPart, invert)
		if len(firstPart) > deg {
			firstPart = firstPart[:deg+1]
		}
		for i := range firstPart {
			if i+curDeg == len(invert) {
				invert = append(invert, 0)
			}
			invert[i+curDeg] = (invert[i+curDeg] + firstPart[i]) % module
		}
		curDeg <<= 1
	}
	return invert, nil
}

func main() {
	scanner := bufio.NewScanner(os.Stdin)
	scanner.Buffer(nil, 1<<30)
	scanner.Split(splitFunc)

	m := ScanInt(scanner)
	n := ScanInt(scanner)
	poly := make([]int, n+1)
	for i := 0; i <= n; i++ {
		poly[i] = ScanInt(scanner)
	}

	fabric = initFabric(module)
	result, err := findInvertPoly(poly, m, &fabric)
	if err != nil {
		fmt.Println(err)
		return
	}
	if len(result) > m {
		result = result[:m]
	}
	for len(result) < m {
		result = append(result, 0)
	}
	for _, coeff := range result {
		fmt.Printf("%d ", coeff%module)
	}
	fmt.Println()
}

File: contest2/f_cpp/f.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#pragma GCC optimize("O3,unroll-loops")
#pragma GCC target("avx2,bmi,bmi2,lzcnt,popcnt")

const int	MODULE = 7340033;
const int	ROOT   = 5;
const int	INVERT = 4404020;
const int	POWER  = 1 << 20;
inline std::vector<int> rev;

inline int getReversed(int num, int log) {
	int left = 0;
	int right = log - 1;
	while (left < right) {
		int bitLeft = (num & (1 << left)) >> left;
		int bitRight = (num & (1 << right)) >> right;
		num ^= (bitLeft << left);
		num ^= (bitRight << left);
		num ^= (bitRight << right);
		num ^= (bitLeft << right);
		++left;
		--right;
	}
	return num;
}

inline int gcd (int a, int b, int& x, int& y) {
	if (a == 0) {
		x = 0; 
		y = 1;
		return b;
	}
	int x_new, y_new;
	int d = gcd(b%a, a, x_new, y_new);
	x = (y_new - ((b / a) * 1ll) * x_new);
	y = x_new;
	return d;
}

inline int invert_number(int num) {
	int x, y;
	gcd(num, MODULE, x, y);
	while (x < 0) {
		x = (x + MODULE) % MODULE;
	}
	return x;
}

inline void fill_rev(int n, int log) {
	for (int i = 0; i < n; ++i) {
		rev[i] = getReversed(i, log);
	}
}

inline void delete_zeros(std::vector<int>& poly) {
	while (poly.size() > 1 && poly[poly.size() - 1] == 0) {
		poly.pop_back();
	}
}

inline void fft(std::vector<int>& poly, int log, bool isInvertFFT) {
	if (poly.size() == 1) {
		return;
	}

	for (int i = 0; i < poly.size(); ++i) {
		int reversed = rev[i];
		if (i < reversed) {
			std::swap(poly[i], poly[reversed]);
		}
	}

	int curOffset = 1;
	for (int i = 0; i < log; ++i) {
		int root = 0;
		if (isInvertFFT) {
			root = INVERT;
		} else {
			root = ROOT;
		}
		for (int j = curOffset << 1; j < POWER; j <<= 1) {
			root = ((root * 1ll) * root) % MODULE;
		}
		for (int j = 0; j < poly.size(); j += curOffset << 1) {
			int curRoot = 1;
			for (int k = 0; k < curOffset; ++k) {
				int tmp1 = poly[k + j];
				int tmp2 = poly[k + j + curOffset];
				poly[k + j] = tmp1 + ((curRoot*1ll)*tmp2)%MODULE;
				if (poly[k + j] > MODULE) {
					poly[k + j] -= MODULE;
				}
				poly[k + j + curOffset] = tmp1 - ((curRoot*1ll)*tmp2)%MODULE;
				if (poly[k + j + curOffset] < 0) {
					poly[k + j + curOffset] += MODULE;
				}
				curRoot = ((curRoot * 1ll) * root) % MODULE;
			}
		}
		curOffset <<= 1;
	}
	if (isInvertFFT) {
		int inverted = invert_number(poly.size());
		for (int i = 0; i < poly.size(); ++i) {
			poly[i] = ((poly[i] * 1ll) * inverted) % MODULE;
		}
	}
}

inline void multiply(std::vector<int>& first_poly, std::vector<int>& second_poly_arg) {	
	if (first_poly.size() == 0 || second_poly_arg.size() == 0) {
		first_poly.resize(1);
		first_poly[0] = 0;
		return;
	}
	int first_size = first_poly.size();
	while (first_size > 1 && first_poly[first_size-1] == 0) {
        --first_size;
	}
	int second_size = second_poly_arg.size();
	while (second_size > 1 && second_poly_arg[second_size-1] == 0) {
		--second_size;
	}

	if (first_poly[first_size-1] == 0 || second_poly_arg[second_size-1] == 0) {
		first_poly.resize(1);
		first_poly[0] = 0;
		return;
	}

	std::vector<int> second_poly = second_poly_arg;
	int min_deg_two = 1;
	int log = 0;
	while (min_deg_two < std::max(first_size, second_size)) {
		++log;
		min_deg_two <<= 1;
	}
	min_deg_two <<= 1;
	++log;
	rev.resize(min_deg_two);
	fill_rev(min_deg_two, log);

	first_poly.resize(min_deg_two);
	second_poly.resize(min_deg_two);

	fft(first_poly, log, false);
	fft(second_poly, log, false);

	for (int i = 0; i < first_poly.size(); ++i) {
		first_poly[i] = ((first_poly[i] * 1ll) * second_poly[i]) % MODULE;
	}

	fft(first_poly, log, true);
}

inline void findInvertPoly(std::vector<int>& invert, std::vector<int>& poly, int deg) {
    if (poly[0] == 0) {
		return;
	}
	if (poly.size() < deg) {
		poly.resize(deg);
	}
	invert.resize(poly.size());
	int	curDeg = 1;
	
	invert[0] = invert_number(poly[0]);
	while (curDeg < deg) {
        std::vector<int> firstPart(poly.begin(), poly.begin() + curDeg);
		std::vector<int> secondPart(poly.begin() + curDeg, poly.end());
		multiply(firstPart, invert);
		while (firstPart.size() > deg) {
			firstPart.pop_back();
		}
		if (curDeg >= firstPart.size()) {
			firstPart.resize(1);
			firstPart[0] = 0;
		} else {
			std::reverse(firstPart.begin(), firstPart.end());
			for (int i = 0; i < curDeg; ++i) {
				firstPart.pop_back();
			}
			std::reverse(firstPart.begin(), firstPart.end());
		}
		multiply(secondPart, invert);
		while (secondPart.size() > deg) {
			secondPart.pop_back();
		}
		for (int i = 0; i < std::max(firstPart.size(), secondPart.size()); ++i) {
			if (i == firstPart.size()) {
                firstPart.push_back(0);
			}
			if (i < firstPart.size() && i < secondPart.size()) {
				firstPart[i] = (-firstPart[i] - secondPart[i] + 2*MODULE) % MODULE;
			} else if (i < secondPart.size()) {
				firstPart[i] = (-secondPart[i] + MODULE) % MODULE;
			} else {
				firstPart[i] = (-firstPart[i] + MODULE) % MODULE;
			}
		}
		multiply(firstPart, invert);
		while (firstPart.size() > deg) {
			firstPart.pop_back();
		}
		for (int i = 0; i < firstPart.size(); ++i) {
			if (i + curDeg == invert.size()) {
                invert.push_back(0);
			}
			invert[i + curDeg] = (invert[i + curDeg] + firstPart[i]) % MODULE;
		}
		curDeg <<= 1;
	}
}

int main() {
    std::ios_base::sync_with_stdio(false);
	std::cin.sync_with_stdio(false);
	std::cout.sync_with_stdio(false);
    std::cin.tie(NULL);
    std::cout.tie(NULL);
    int m;
    std::cin >> m;
    int n;
    std::cin >> n;
    std::vector<int> poly(n + 1);
	for (int i = 0; i <= n; ++i) {
        std::cin >> poly[i];
	}
	std::vector<int> result;
	findInvertPoly(result, poly, m);
	if (result.size() == 0) {
		std::cout << "The ears of a dead donkey" << '\n';
	} else {
		result.resize(m);
		for (int i = 0; i < result.size(); ++i) {
			std::cout << (result[i] % MODULE) << " ";
		}
		std::cout << '\n';
	}
}